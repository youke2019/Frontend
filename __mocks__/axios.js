const axios = jest.genMockFromModule('axios');

let mockFiles = Object.create(null);
function post(url,data) {
    const promise = new Promise(function (resolve, reject) {
        if (
            data.answer_content == 'error'
        ) {
            reject("error")
        } else {
            resolve(data)
        }
    })
    return promise
}

function get(data) {
    const promise = new Promise(function (resolve, reject) {
        if (false) {
            resolve(data);
        } else {
            reject("error");
        }
    })
    return promise
}

axios.post = post;
axios.get = get;

module.exports = axios;