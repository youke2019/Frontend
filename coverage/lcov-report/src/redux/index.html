<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/redux</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="..\..\prettify.css" />
    <link rel="stylesheet" href="..\..\base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(..\..\sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="..\..\index.html">All files</a> src/redux
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">44.05% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>37/84</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">36% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>9/25</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">25% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>5/20</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">50% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>36/72</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line low'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file medium" data-value="actions.js"><a href="actions.js.html">actions.js</a></td>
	<td data-value="69.7" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 69%;"></div><div class="cover-empty" style="width:31%;"></div></div></td>
	<td data-value="69.7" class="pct medium">69.7%</td>
	<td data-value="33" class="abs medium">23/33</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="9.09" class="pct low">9.09%</td>
	<td data-value="11" class="abs low">1/11</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="22" class="abs high">22/22</td>
	</tr>

<tr>
	<td class="file low" data-value="reducers.js"><a href="reducers.js.html">reducers.js</a></td>
	<td data-value="24.49" class="pic low"><div class="chart"><div class="cover-fill" style="width: 24%;"></div><div class="cover-empty" style="width:76%;"></div></div></td>
	<td data-value="24.49" class="pct low">24.49%</td>
	<td data-value="49" class="abs low">12/49</td>
	<td data-value="36" class="pct low">36%</td>
	<td data-value="25" class="abs low">9/25</td>
	<td data-value="44.44" class="pct low">44.44%</td>
	<td data-value="9" class="abs low">4/9</td>
	<td data-value="25" class="pct low">25%</td>
	<td data-value="48" class="abs low">12/48</td>
	</tr>

<tr>
	<td class="file high" data-value="state.js"><a href="state.js.html">state.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Thu Aug 01 2019 01:16:28 GMT+0800 (GMT+08:00)
</div>
</div>
<script src="..\..\prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="..\..\sorter.js"></script>
<script src="..\..\block-navigation.js"></script>
</body>
</html>
