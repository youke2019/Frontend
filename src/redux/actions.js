export const UPDATE_USER_INFO = 'UPDATEUSERINFO'
export const LOAD_USER_INFO = 'LOADUSERINFO'

export const updateUserInfo = (data) => ({
    type:UPDATE_USER_INFO,
    data
})

export const loadUserInfo = (data) => ({
    type:LOAD_USER_INFO,
    data
})

