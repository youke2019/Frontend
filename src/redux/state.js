const initialState = {
    user_info: null,
    course_list: null,
    login_ready:false,
}

export default initialState