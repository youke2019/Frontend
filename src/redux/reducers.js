import { combineReducers } from 'redux'
import {
    UPDATE_USER_INFO,
    LOAD_USER_INFO,
    CLEAR_USER_INFO,
    SEARCH_COURSES,
} from "./actions.js"
import initialState from "./state";
import axios from 'axios'


function userInfoReducer(state=initialState.user_info, action) {
    switch (action.type) {
        case UPDATE_USER_INFO:
            storage.save({
                key: 'user',
                data: action.data
            })
            state = action.data
            return state
        case LOAD_USER_INFO:
            state = action.data
            return state
        case CLEAR_USER_INFO:
            state = null
            return state
        default:
            return state
    }
}

function courseReducer(state=initialState.course_list, action) {
    switch (action.type) {
        case SEARCH_COURSES:
            state = action.data
            return state
        default:
            return state
    }
}

export const combinedReducer = combineReducers({
    user_info : userInfoReducer,
    course_list: courseReducer,
})