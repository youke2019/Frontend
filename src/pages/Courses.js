import React from "react";
import {
    StyleSheet,
    View,
    Image,
    ImageBackground,
    Text, TouchableWithoutFeedback,
} from "react-native";
import {Button} from "react-native-elements"
import Drawer from '../components/Drawer'
import SearchBox from '../components/SearchBox'
import CourseList from '../components/CourseList'
import axios from "axios";
import {searchCourses} from "../redux/actions";
import {connect} from "react-redux";

const mapStateToProps = state => {return{}}

const mapDispatchToProps = dispatch => {
    return {
        searchCourses: (data) => {
            dispatch(searchCourses(data))
        }
    }
}

class Courses extends React.Component {
    state = {
        filterVisible:false,
        filterList: {
            学分:[],
            通识类型:[],
            上课时间:[],
            开设学院:null,
            上课地点:null,
        },
        firstEnter: true,
        tagList: [{type:'学分',tag:[]},{type:'通识类型',tag:[]},{type:'上课时间',tag:[]},{type:'上课地点',tag:[]},{type:'开设学院',tag:[]}]
    }

    filterNotEmpty = () => {
        if (this.state.filterList['学分'].length > 0)
            return true
        if (this.state.filterList['通识类型'].length > 0)
            return true
        if (this.state.filterList['上课时间'].length > 0)
            return true
        if (this.state.filterList['开设学院'] != null)
            return true
        if (this.state.filterList['上课地点'] != null)
            return true

        return false
    }

    search = (keyword) => {
        const {filterList} = this.state

        if (keyword != '' || this.filterNotEmpty()) {
            let data = {
                course_name:keyword,
                course_credits: [],
                general_types: [],
                weekdays: [],
                building: null,
                dept_name: null,
            }

            this.setState({
                firstEnter: false,
            })

            for (let item of this.state.filterList['学分'])
                data.course_credits.push(item.value)
            if (data.course_credits.length == 0)
                delete data.course_credits

            for (let item of this.state.filterList['通识类型'])
                data.general_types.push(item.value)
            if (data.general_types.length == 0)
                delete data.general_types

            for (let item of this.state.filterList['上课时间'])
                data.weekdays.push(item.value)
            if (data.weekdays.length == 0)
                delete data.weekdays


            if (filterList['开设学院'] == null)
                delete data.dept_name
            else
                data.dept_name = filterList['开设学院'].name

            if (filterList['上课地点'] == null)
                delete data.building
            else
                data.building = filterList['上课地点'].name

            axios.post(baseUrl+'/courses/search',data).then((response) => {
                this.props.searchCourses(response.data)
            }).catch(err => {
                console.log(err)
            })
        }
    }

    onFilterReturn = (filterList) => {
        if (Object.getOwnPropertyNames(filterList).length != 0)
        {
            this.setState({
                filterList: filterList,
                filterVisible: false
            })

            let tagList = [{type:'学分',tag:[]},{type:'通识类型',tag:[]},{type:'上课时间',tag:[]},{type:'上课地点',tag:[]},{type:'开设学院',tag:[]}]
            for (let i=0;i<filterList['学分'].length;i++)
                tagList[0].tag.push({
                    index: i,
                    name: filterList['学分'][i].name
                })

            for (let i=0;i<filterList['通识类型'].length;i++)
                tagList[1].tag.push({
                    index: i,
                    name: filterList['通识类型'][i].name
                })

            for (let i=0;i<filterList['上课时间'].length;i++)
                tagList[2].tag.push({
                    index: i,
                    name: filterList['上课时间'][i].name
                })

            if (filterList['上课地点'] != null)
                tagList[3].tag.push({
                    name: filterList['上课地点'].name
                })

            if (filterList['开设学院'] != null)
                tagList[4].tag.push({
                    name: filterList['开设学院'].name
                })

            this.setState({
                tagList : tagList
            })
        } else {
            this.setState({
                filterList: this.state.filterList,
                filterVisible: false
            })
        }

    }

    showFilter = () => {
        this.setState({
            filterVisible:true
        })
    }

    deleteTag = (type,typeIndex,tag) => {
        let filterList = this.state.filterList
        let tagList = this.state.tagList

        if (typeIndex > 2)
            filterList[type] = null
        else {
            for (let i=0;i<filterList[type].length;i++)
                if (filterList[type][i].name == tag.name)
                    filterList[type].splice(i,1)
        }

        for (let i=0;i<tagList[typeIndex].tag.length;i++)
            if (tagList[typeIndex].tag[i].name == tag.name)
                tagList[typeIndex].tag.splice(i,1)

        this.setState({
            filterList: filterList,
            tagList: tagList
        })
    }

    render() {
        const {
            firstEnter,
            filterVisible,
            filterList,
            tagList
        } = this.state

        return (
            <View style={{flex:1}}>
                {
                    firstEnter?
                        <View style={styles.initial_container}>
                            <Image
                                resizeMode='contain'
                                style={styles.logo}
                                source={{uri:'logo'}}
                            />
                            <SearchBox
                                onPress={(keyword) => this.search(keyword)}
                                iconImage = 'search'
                                buttonTitle = '搜索'
                            />
                        </View>
                        :
                        <ImageBackground
                            style={{flex:1}}
                            imageStyle={{resizeMode: 'cover'}}
                            source={{uri:'course_background'}}
                        >
                            <View style={styles.container}>
                                <View style={styles.search_container}>
                                    <View style={{flex:1}}>
                                        <Button
                                            icon={<Image
                                                style={styles.filter}
                                                source={{uri:'filter'}}
                                            />}
                                            containerStyle={styles.button}
                                            titleStyle={styles.text}
                                            onPress={this.showFilter}
                                            type="clear"
                                            title="筛选"
                                        />
                                    </View>
                                    <View style={{flex:5}}>
                                        <SearchBox
                                            onPress={(keyword) => this.search(keyword)}
                                            iconImage = 'search'
                                            buttonTitle = '搜索'
                                        />
                                    </View>
                                </View>
                                <View style={styles.tag_container}>
                                    {
                                        tagList.map((item,index) => {
                                            return (
                                                <View style={styles.tag_list} key={index}>
                                                    {
                                                        item.tag.length > 0 ?
                                                            <Text style={styles.tag_type_text}>{item.type}:</Text>
                                                            :
                                                            null
                                                    }
                                                    {
                                                        item.tag.map((tag) => {
                                                            return (
                                                                <Button
                                                                    key={tag.name}
                                                                    title={tag.name}
                                                                    type="clear"
                                                                    onPress={()=>this.deleteTag(item.type,index,tag)}
                                                                    icon={<Image source={{uri:'cancel'}} style={{width:16,height:16}}/>}
                                                                    iconRight={true}
                                                                    containerStyle={styles.tag_button_container}
                                                                    buttonStyle={styles.tag_button}
                                                                    titleStyle={styles.tag_title}
                                                                    TouchableComponent={TouchableWithoutFeedback}
                                                                />
                                                            )
                                                        })
                                                    }
                                                </View>
                                            )
                                        })
                                    }
                                </View>
                            </View>
                            <CourseList navigation = {this.props.navigation}/>
                            <Drawer
                                visible={filterVisible}
                                onFilterReturn = {this.onFilterReturn}
                                list={filterList}
                            />
                        </ImageBackground>
                }
            </View>
        )
    }
}

const styles = StyleSheet.create({
    initial_container:{
        flex:1,
        justifyContent: 'center',
        alignItems: 'center',
    },
    logo:{
        width: 600,
        height: 80,
    },
    container:{
        paddingTop:15,
    },
    search_container:{
        flexDirection: 'row',
        alignItems: 'center',
    },
    button_container:{
        justifyContent: 'center',
        alignItems: 'center',
    },
    filter_container:{
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
    },
    filter:{
        height: 15,
        width: 15,
    },
    text:{
        color: 'black',
        fontSize: 14,
    },
    tag_container:{
        flexDirection: 'row',
        flexWrap: 'wrap',
    },
    tag_list:{
        flexDirection: 'row',
        alignItems: 'center',
    },
    tag_type_text:{
        paddingHorizontal: 5,
    },
    tag_button_container:{
        padding:5,
    },
    tag_button:{
        borderColor: '#FDAF26',
        borderWidth: 1,
        borderRadius: 30,
        padding: 5,
    },
    tag_title:{
        color: '#ff961e',
        fontSize: 15,
    }
})

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Courses)