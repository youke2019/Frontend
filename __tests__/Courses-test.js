import React from 'react';
import renderer from 'react-test-renderer';
import Courses from '../src/pages/Courses'
import { createStore } from 'redux'
import { combinedReducer } from '../src/redux/reducers'
import { Provider } from 'react-redux'
import Adapter from 'enzyme-adapter-react-16'
import {shallow,mount,configure} from 'enzyme'
import axios from 'axios'

configure({adapter: new Adapter()})

const testInitialState = {
    user_info: null,
    course_list: JSON.parse("[{\"course_id\":\"CV440\",\"course_name\":\"土木工程应用软件\",\"course_hours\":32,\"course_credits\":2,\"general\":false,\"general_type\":\"\",\"course_deptname\":\"船舶海洋与建筑工程学院\"},{\"course_id\":\"EE312\",\"course_name\":\"软件工程（C类）\",\"course_hours\":34,\"course_credits\":2,\"general\":false,\"general_type\":\"\",\"course_deptname\":\"电子信息与电气工程学院\"},{\"course_id\":\"EM403\",\"course_name\":\"大型工程软件应用\",\"course_hours\":48,\"course_credits\":3,\"general\":false,\"general_type\":\"\",\"course_deptname\":\"船舶海洋与建筑工程学院\"},{\"course_id\":\"IM414\",\"course_name\":\"计算机硬件与软件原理\",\"course_hours\":48,\"course_credits\":3,\"general\":false,\"general_type\":\"\",\"course_deptname\":\"安泰经济与管理学院\"},{\"course_id\":\"IS415\",\"course_name\":\"系统软件课程设计\",\"course_hours\":34,\"course_credits\":2,\"general\":false,\"general_type\":\"\",\"course_deptname\":\"电子信息与电气工程学院\"},{\"course_id\":\"SE100\",\"course_name\":\"软件工程创新实践基础\",\"course_hours\":32,\"course_credits\":2,\"general\":false,\"general_type\":\"\",\"course_deptname\":\"电子信息与电气工程学院\"},{\"course_id\":\"SE300\",\"course_name\":\"软件工程\",\"course_hours\":56,\"course_credits\":4,\"general\":false,\"general_type\":\"\",\"course_deptname\":\"电子信息与电气工程学院\"},{\"course_id\":\"SE407\",\"course_name\":\"软件工程\",\"course_hours\":16,\"course_credits\":1,\"general\":false,\"general_type\":\"\",\"course_deptname\":\"电子信息与电气工程学院\"},{\"course_id\":\"SE419\",\"course_name\":\"企业软件质量保证\",\"course_hours\":16,\"course_credits\":1,\"general\":false,\"general_type\":\"\",\"course_deptname\":\"电子信息与电气工程学院\"},{\"course_id\":\"SE422\",\"course_name\":\"企业软件过程与管理\",\"course_hours\":16,\"course_credits\":1,\"general\":false,\"general_type\":\"\",\"course_deptname\":\"电子信息与电气工程学院\"},{\"course_id\":\"XP401\",\"course_name\":\"生产实习（软件工程）\",\"course_hours\":256,\"course_credits\":16,\"general\":false,\"general_type\":\"\",\"course_deptname\":\"电子信息与电气工程学院\"}]"),
}


const store = createStore(combinedReducer,testInitialState)

test('renders course page correctly', () => {
    const tree = renderer.create(
        <Provider store={store}>
            <Courses navigation = { null} />
        </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});

describe('functionality', () => {
    let wrapper = shallow(<Courses store={store}/>).dive().dive()

    it('filterVisible functions', () => {

        console.log(wrapper.instance())
        wrapper.instance().showFilter()
        expect(wrapper.state('filterVisible')).toBe(true)
        wrapper.instance().onBackdropPress()
        expect(wrapper.state('filterVisible')).toBe(false)
    })

    it('filterList functions', () => {
        expect(wrapper.instance().filterNotEmpty()).toBe(false)
        wrapper.instance().updateFilter({
            学分:['1','2']
        })
        console.log(wrapper.state('filterList')['学分'])
        expect(wrapper.state('filterList')['学分']).toStrictEqual(['1','2'])
        expect(wrapper.instance().filterNotEmpty()).toBe(true)
    })

    it('search function', (done) => {
        jest.mock('axios')
        global.baseUrl = 'baseUrl'
        wrapper.instance().search('花')
        const promise = new Promise(
            function (resolve, reject)
            {resolve('success')})
        axios.post()
            .then(data=>{
                console.log(wrapper.props())
                done()
            })
            .catch(data=>{done()})

    })
})
