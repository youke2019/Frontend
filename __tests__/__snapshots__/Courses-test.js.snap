// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders course page correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Image
      resizeMode="contain"
      source={
        Object {
          "uri": "logo",
        }
      }
      style={
        Object {
          "height": 80,
          "width": 600,
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 80,
            "justifyContent": "center",
          },
          Object {},
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "borderRadius": 20,
              "flex": 7,
              "flexDirection": "row",
              "height": 40,
              "justifyContent": "center",
              "marginLeft": 12,
            },
            Object {
              "backgroundColor": "white",
              "borderWidth": 1,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
              "paddingLeft": 6,
            }
          }
        >
          <Image
            source={
              Object {
                "uri": "search",
              }
            }
            style={
              Object {
                "height": 16,
                "width": 16,
              }
            }
          />
        </View>
        <TextInput
          allowFontScaling={true}
          onChangeText={[Function]}
          rejectResponderTermination={true}
          style={
            Object {
              "flex": 6,
              "fontSize": 15,
              "padding": 2,
            }
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
      <View
        style={
          Object {
            "flex": 2,
            "padding": 6,
          }
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontFamily": "字魂95号-手刻宋",
                "fontSize": 18,
              }
            }
          >
            搜索
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
